presc/copies/sampling.py

```diff
@@ -557,3 +563,115 @@ def labeling(X, original_classifier, label_col="class"):
+):
+    """Generate balanced synthetic data using any sampling function.
+
+    This function will attempt to obtain a balanced dataset with non
+    balancing samplers by generating the same number of samples for all
+    classes, unless it reaches the maximum number of iterations. To use
+    within the ClassifierCopy class, the `enforce_balance` must be set
+    to True.
+
+    Note that the algorithm needs to find at least one sample of a different
+    class in order detect that class and keep iterating through the batch
+    generation of samples to try to get them all. Therefore, it is not
+    guaranteed that it will find all the classes and successfully balance the
+    synthetic dataset in extreme cases of imbalance. However, the batch size
+    `nbatch` can be set to a larger number if we suspect that is the case, and
+    this will increase the probability to find at least a sample of a different
+    class in the first round. Thereafter, if the algorithm is already iterating
+    to find a minoritary class, it is more likely that other classes that occupy
+    a very small hypervolume will show up as well.
+
+    Parameters
+    ----------
+    feature_parameters : dict of dicts
+        A dictionary with an entry per dataset feature (dictionary keys should
+        be the feature names), where each feature entry must contain a
+        nested dictionary with its categories and their fraction. The key for
+        the nested dictionary of categories should be "categories", and the keys
+        for the fractions should be the category name.
+    numerical_sampling : function
+        Any of the non balancing numerical sampling functions defined in PRESC:
+        `grid_sampling`, `uniform_sampling`, `normal_sampling`...
+    original_classifier : sklearn-type classifier
+        Original ML classifier used to generate the synthetic data.
+    nsamples : int
+        Number of samples to generate.
+    max_iter : int
+        The maximum number of iterations generating batches to attempt to obtain
+        the samples per class specified in `nsamplesxclass`.
+    nbatch: int
+        Number of tentative samples to generate in each batch.
+    label_col : str
+        Name of the label column.
+    random_state : int
+        Random seed used to generate the sampling data.
+    verbose : bool
+        If True the sampler prints information about each batch.
+
+    Returns
+    -------
+    pandas DataFrame
+        Dataset with a generated sampling following the specified numerical
+        sampling distribution for the numerical features and the discrete
+        distribution for the categorical features, following the feature space
+        characterized by the `feature_parameters`, where the function has
+        tried to balance the samples for each class.
+    """
+    if random_state is not None:
+        np.random.seed(seed=random_state)
+
+    nsamplesxclass = int(nsamples / len(feature_parameters))
```

Comment:
To get the target number of samples per class, this divides the target number of samples by the number of features. I'm not sure this is what we want since the number of classes doesn't necessarily depend on the features. Could we pass in `nsamplesxclass` as a parameter?